docs/source/examples/pyquil_demo.myst

```diff
@@ -0,0 +1,174 @@
+---
+jupytext:
+  text_representation:
+    extension: .myst
+    format_name: myst
+    format_version: '0.8'
+    jupytext_version: '1.4.1'
+kernelspec:
+  display_name: Python 3
+  language: python
+  name: python3
+---
+
+# Zero Noise Extrapolation with pyQuil and parametric compilation
+In this example we demonstrate how to use Zero Noise Extrapolation (ZNE) in Mitiq with a parametrically compiled program in pyQuil. Parametric compilation saves time by compiling the program ansatz once, instead of compiling each time the parameters are updated. In pyQuil 2.28.1, a `memory_map` argument is used to substitute in values for previously-declared Quil variables in the pre-compiled executable. 
+
+When adding the noise mitigation function `zne.execute_w_zne` the `memory map` is still passed to the executor function and the circuit parameters are updated without the need for repeated compilation. 
+
+The PyQuil parametric compilation example shown here is adapted from https://pyquil-docs.rigetti.com/en/v2.28.1/migration3-declare.html
+
+
+```{code-cell}
+import numpy as np
+from pyquil import get_qc, Program
+from pyquil.gates import RX, RY, H, CNOT, MEASURE
+import mitiq
+from mitiq import zne
+from scipy import optimize
+```
++++
+
+Use the get_qc command to initialize the simulated noisy device where the PyQuil program will run.
+
+```{code-cell}
+# initialize the quantum device
+qc = get_qc("2q-noisy-qvm")
+```
++++
+
+Set up the quantum circuit
+
+```{code-cell}
+def ansatz():
+    program = Program()
+    theta = program.declare("theta", memory_type="REAL")
+    ro = program.declare("ro", memory_type="BIT", memory_size=1)
+    program += RX(theta, 0)
+    program += RY(theta, 0) 
+    program += RX(theta, 1)
+    program += RY(theta, 1)
+    program += H(0)
+    program += CNOT(1,0)
+    program += H(0)
+    program += MEASURE(0, ro[0])
+    return program
```

Comment:
I think it's better to inline this code (i.e., define the program not inside a function). This emphasizes that you only build the program once and use parametric compilation. I only see one `program = ansatz()` line, so this should be a very easy change.

Reply:
I agree- code is updated.